### PR TITLE
replace MICROPY_PORT_ROOT_POINTERS with MP_REGISTER_ROOT_PONTER

### PR DESCRIPTION
**Pull request by dlech**
Based on the discussion in #8278, this follows the precedent set by #8566.

This creates a new automatically generated header file that contains root pointers and associated tooling for registering a pointer in the file relevant to the code where it is used instead of having all in `mpconfigport.h` configuration files. All root pointers are then converted to use this new macro. Finally the old `MICROPY_BOARD_ROOT_POINTERS` is removed.

### CODE REVIEW

**Comment by dlech** [comment]
There are still a few kinks to work out. Are root pointers needed when building `mboot` on STM32?

And there are still more root pointers to be converted, I've only done the port-specific ones so far.

**Comment by dlech** [comment]
Looks like I got the kinks worked out.

**Comment by robert-hh** [comment]
That's quite a change at many places, requiring a huge amount of re-testing. Missing root pointer definitions do not show up at compile time, and usually not at the first test run. They cause problems when objects are wiped away at some time by GC.
It would be nice to have some of the registrations configurable, especially for the readline_hist[n] number, which can be made larger for ports/boards wit a lot of RAM.

**Comment by dlech** [comment]
Thanks for having a look. 

> Missing root pointer definitions do not show up at compile time, and usually not at the first test run.

I'm not sure I understand this concern. If a `MP_REGISTER_ROOT_POINTER()` is missing then you get compiler error like this:

```
In file included from ../../py/mpstate.h:33,
                 from input.c:33:
input.c: In function ‘prompt_write_history’:
../../py/mpstate.h:329:40: error: ‘mp_state_vm_t’ {aka ‘struct _mp_state_vm_t’} has no member named ‘readline_hist’
  329 | #define MP_STATE_VM(x) (mp_state_ctx.vm.x)
      |                                        ^
../../py/misc.h:122:34: note: in definition of macro ‘MP_ARRAY_SIZE’
  122 | #define MP_ARRAY_SIZE(a) (sizeof(a) / sizeof((a)[0]))
      |                                  ^
./mpconfigport.h:291:23: note: in expansion of macro ‘MP_STATE_VM’
  291 | #define MP_STATE_PORT MP_STATE_VM
      |                       ^~~~~~~~~~~
input.c:111:40: note: in expansion of macro ‘MP_STATE_PORT’
  111 |             for (int i = MP_ARRAY_SIZE(MP_STATE_PORT(readline_hist)) - 1; i >= 0; i--) {
      |                                     
```

This is is not any different than before this change if you forget to add something to `MICROPY_PORT_ROOT_POINTERS `.

> It would be nice to have some of the registrations configurable, especially for the readline_hist[n] number, which can be made larger for ports/boards wit a lot of RAM.

Yes. In fact the second patch here contains a new `MICROPY_READLINE_HISTORY_SIZE ` config option exactly for this.

**Comment by jimmo** [comment]
Fantastic work @dlech, thank you!!

Yes, I agree, unlike the move to `MP_REGISTER_MODULE`, the compiler should catch any problems here.

I suspect at some point we will need to probably need to address the fact that a script called `makeqstrdefs.py` is doing a lot more than just qstr processing, but that can be done later.

When I said "and maybe root pointers?" in #8566 I didn't consider that this would extend to the ones in mpstate.h (i.e. non-port specific ones), but honestly I think this is a good improvement.

**Comment by dpgeorge** [comment]
I also shared @robert-hh 's concern about accidentally missing something.  But very good to know that the compiler will pick it up.  As long as you remember to access the root pointer using the `MP_STATE_VM()` macro (which is what you needed to do previously) it will be safe.

I guess the only other general concern is that we are really going all-in on the preprocessing stage, it's a big part of the build process now.  I haven't seen anyone raise a practical issue about this yet, so that's encouraging.

File: extmod/nimble/modbluetooth_nimble.c [hunk 1]
```diff
@@ -1966,4 +1966,7 @@ STATIC int ble_secret_store_delete(int obj_type, const union ble_store_key *key)
 
 #endif // MICROPY_PY_BLUETOOTH_ENABLE_PAIRING_BONDING
 
+MP_REGISTER_ROOT_POINTER(struct _mp_bluetooth_nimble_malloc_t *bluetooth_nimble_memory);
```
Comment (dpgeorge) [inline extmod/nimble/modbluetooth_nimble.c:1969]: I think this line should be wrapped in `#if !MICROPY_BLUETOOTH_NIMBLE_BINDINGS_ONLY`

File: ports/esp8266/machine_uart.c [hunk 2]
```diff
@@ -329,3 +329,5 @@ const mp_obj_type_t pyb_uart_type = {
     .protocol = &uart_stream_p,
     .locals_dict = (mp_obj_dict_t *)&pyb_uart_locals_dict,
 };
+
+MP_REGISTER_ROOT_POINTER(byte * uart0_rxbuf);
```
Comment (dpgeorge) [inline ports/esp8266/machine_uart.c:333]: Can you remove the space after `*` for formatting consistency?

File: py/make_root_pointers.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
+"""
```
Comment (dpgeorge) [inline py/make_root_pointers.py:1]: Please rename this file to `makerootpointers.py` for consistency with other scripts.

File: py/modsys.c [hunk 4]
```diff
@@ -286,4 +286,24 @@ const mp_obj_module_t mp_module_sys = {
...
 MP_REGISTER_MODULE(MP_QSTR_usys, mp_module_sys);
 
+// If MICROPY_PY_SYS_PATH_ARGV_DEFAULTS is not enabled then these two lists
+// must be initialised after the call to mp_init.
+MP_REGISTER_ROOT_POINTER(mp_obj_list_t mp_sys_path_obj);
+MP_REGISTER_ROOT_POINTER(mp_obj_list_t mp_sys_argv_obj);
+
+#if MICROPY_PY_SYS_EXC_INFO
+// current exception being handled, for sys.exc_info()
+MP_REGISTER_ROOT_POINTER(mp_obj_base_t * cur_exception);
```
Comment (dpgeorge) [inline py/modsys.c:296]: Remove space after `*` ?

File: shared/runtime/pyexec.c [hunk 5]
```diff
@@ -488,6 +488,8 @@ int pyexec_event_repl_process_char(int c) {
     return res;
 }
 
+MP_REGISTER_ROOT_POINTER(vstr_t * repl_line);
```
Comment (dpgeorge) [inline shared/runtime/pyexec.c:491]: Remove space after `*`?

**Comment by dpgeorge** [comment]
Thanks @dlech this looks really good.

There are 64 commits.  What do you think about squashing those related to a given port, eg all stm32 changes are in one commit (maybe excluding extmod changes like bluetooth which can stay separate)?  At least for the cc3200 ones they should squash without any problem, and it might be a bit cleaner that way (in the git history).

File: ports/esp8266/machine_uart.c (hunk 2 above)
Comment (dlech) [reply ports/esp8266/machine_uart.c:333]: The CI code format check fails without the space.

File: py/make_root_pointers.py (hunk 3 above)
Comment (dlech) [reply py/make_root_pointers.py:1]: Normally, I am in favor of consistency, but I have a really hard time reading words that are squished together like this. My brain moves/adds/removes letters until it finds a word that it recognizes, but it is rarely what is actually written. So, I was hoping to start a new trend of being more readable.

File: ports/esp8266/machine_uart.c (hunk 2 above)
Comment (dlech) [reply ports/esp8266/machine_uart.c:333]: e.g. https://github.com/micropython/micropython/runs/7156973079?check_suite_focus=true

File: py/make_root_pointers.py (hunk 3 above)
Comment (dpgeorge) [reply py/make_root_pointers.py:1]: OK, let's keep it as `make_root_pointers.py`.

File: ports/esp8266/machine_uart.c (hunk 2 above)
Comment (dpgeorge) [reply ports/esp8266/machine_uart.c:333]: OK

File: extmod/nimble/modbluetooth_nimble.c (hunk 1 above)
Comment (dlech) [reply extmod/nimble/modbluetooth_nimble.c:1969]: done

**Comment by dlech** [comment]
This has been rebased and port-specific commits have been squashed.

**Comment by dpgeorge** [comment]
Thanks for updating and squashing the port commits.

Merged in fc3d7ae11be11a7f05709ebfd439061fce9ee555 through a1ef5ac65db8c7509bef5d3db9ba6de9fbb4ef40

This is a really great clean-up and simplification for root pointer registration.  It has reduced RAM usage by a few words (pointers) on some builds when certain features are disabled but the root pointer was not conditional (eg HID data pointer on stm32 when HID is disable).

**Comment by robert-hh** [comment]
Trying to pull that change into the open PR for SAMD is quite a nightmare.

**Comment by dlech** [comment]
Sorry to hear that. Fixing bitrot is not fun. What specifically made it do difficult?

**Comment by robert-hh** [comment]
Finally it was not that complicated. It was more the problem that the SAMD PR has so many commits, and ~16 of them change mpconfigport.h. But keeping the changes and making a final change at the end made it easy.

**Pull request closed without merging**